test/library/standard/Map/types/ClassWithMapWithRecord.chpl

```diff
@@ -1,46 +0,0 @@
-use IO, Map;
```

Comment:
Does this test now work?  Why delete it instead of leaving it to continue to check the record case?

Reply:
This test does not work now, but it now has a different failure mode:

```chpl
asd.chpl:21: error: cannot assign to const variable
asd.chpl:41: error: cannot assign to const variable
```

And, looking at the test, I don't really understand what it was trying to do since it is trying to modify a `const`, so it wasn't obvious to me how we would adjust it.

So, since it is no longer serving the function it was originally created for, it made more sense to me to just remove it. Do you disagree with that?

Reply:
I think there's a little more going on here.

1. I think the `const` errors are about the new "this accessor" returning a `const ref`, not due to the `const myC` itself. 

The error refers to assigning to a const, not modifying a const, so it's `myC.files[filename]` (on line 21) that it's finding is const.  For a `const c: C = new C(); c.x = 1;` is valid.  It's just that `c` itself can't be reassigned.

When I make this change on main,

```diff
--- a/test/library/standard/Map/types/ClassWithMapWithRecord.chpl
+++ b/test/library/standard/Map/types/ClassWithMapWithRecord.chpl
@@ -5,8 +5,9 @@ const filename = "foo";
 // a simplified example
 
 record MyRecord {
-  var de: unmanaged object;
-// this works: var de = new unmanaged object();
+  //var de: unmanaged object?;
+// this works:
+  var de = new unmanaged object();
 }
 
 class MyClass {
@@ -29,7 +30,8 @@ delete myC;
 // the original recursive case by @cassella
 
 record recordthatcontainsadirent {
-  var de: unmanaged dirent;
+  //  var de: unmanaged dirent;
+  var de = new unmanaged dirent();
 }
 
 class dirent {
```

it compiles and its output matches the .good file.  So the const failure with the PR just seems like the PR resulting in the `const ref` return being chosen instead of `ref`?  Is that behavior in this case deliberate?  (I haven't really tried to understand the PR.)


2. What happens with the PR if the map entries are made default initializable?

 Changing just `unmanaged object` to `unmanaged object?` (and `unmanaged dirent?` in the second part) still results in the internal error on main.

If the change to a `const ref` `this` here was related to the not-default-initializable map entries, does making this change likewise allow continuing to hit the internal error on the PR branch?


3. Even if this change to the standard Map module means the internal error can't be hit through the Map module any more, the compiler error is still there, so there should still be a future for it.

I'd think the PR should find another way for a future to hit it, even if the future no longer belongs in `test/library/standard/Map`.  The least effort / least elegant way that occurs to me would be to provide the future with its own clone of Map on main today.

Reply:
> seems like the PR resulting in the const ref return being chosen instead of ref? Is that behavior in this case deliberate? 

This was not deliberate and was addressed in response to your other comment. The `where` clause has been removed so we correctly get a `ref` and not a `const ref` in that case.

> If the change to a const ref this here was related to the not-default-initializable map entries, does making this change likewise allow continuing to hit the internal error on the PR branch?

I was not still seeing the error with change on this PR branch.

> I'd think the PR should find another way for a future to hit it, even if the future no longer belongs in test/library/standard/Map. The least effort / least elegant way that occurs to me would be to provide the future with its own clone of Map on main today.

I've updated the PR with a trimmed down version of map that contains just the `this` accessors to preserve the futures hitting of the internal error.

Thanks again for the thorough review!